README.md
<table border="0">
  <tr>
    <td align="left" valign="middle">
    <h1>Wi-SUN Border Router Container</h1>
  </td>
  <td align="left" valign="middle">
    <a href="https://wi-sun.org/">
      <img src="pics/Wi-SUN-logo.png" title="Wi-SUN" alt="Wi-SUN Logo" width="300"/>
    </a>
  </td>
  </tr>
</table>

In a Wi-SUN network, the Border Router (BR) is in charge of managing the network
(authentication, routing, etc.) and providing internet connectivity to the
network.

This Docker image is provided to simplify the deployment and network
configuration. It is developed for running on a Raspberry Pi and Raspberry Pi OS
(a port of Debian for Raspberry Pi), but it should work on any Linux host.

This implementation relies on an external Silicon Labs EFR32xG12 flashed with a
dedicated Wi-SUN RCP (Radio Coprocessor) firmware. To this end, the EFR32 is
connected to the host through USB. Using this connection, the Docker image sees
the EFR32 device as a serial (UART) connection. This document describes how to
set up the system shown below.

![Linux border router](pics/linux_br_image.png)

The use of a network with IPv6 connectivity is encouraged. If IPv6 is not
available, you can use "local" or "site-local" modes (these modes have some
pitfalls, see below).

# Prerequisites

The document requires the following hardware prerequisites:
  * A Raspberry Pi 3 Model B+ or newer
  * An SD card of at least 8 GB
  * A Silicon Labs WTSK mainboard
  * An EFR32MG12 or EFR32FG12 radio board supporting a sub-GHz band

To flash Raspberry Pi OS on the SD card, follow [the steps described by the
Raspberry Pi Foundation.][1]

It is recommended to update and upgrade your OS. To do so, connect the Raspberry
Pi to internet via Ethernet or Wi-Fi and open a console or SSH (Secure Shell)
session.

Update the package database:

    sudo apt-get update

Upgrade installed packages:

    sudo apt-get upgrade

[1]: https://www.raspberrypi.org/software/

# Build the Docker Image

For the time being, the Docker image has to be built. To build and run Docker
images, first install the `docker` application.

    sudo apt-get install docker.io

> This image is compatible with Docker 20 and newer (Debian 11 or newer).

Make sure that your current user is allowed to run Docker:

    sudo usermod -aG docker pi

> You have to log out and back in for this to take effect.

You should now be able to clone the repository on your Raspberry Pi:

    git clone git@github.com:SiliconLabs/wisun-br-linux-docker.git

Enter the `wisun-br-linux-docker` directory:

    cd wisun-br-linux-docker

Build the image with:

    DOCKER_BUILDKIT=1 docker build --build-arg "GIT_DESCRIBE=$(git describe --tag --match v*)" -t wisun-img .

You may save memory by removing the build environment and only keeping the final
image:

    docker image prune

If you have an IPv6 network, create a macvlan interface to allow the Docker
image to access it (replace `eth0` by the name of your physical network
interface):

    sudo docker network create -d macvlan -o parent=eth0 wisun-net

> Details on Docker and macvlan can be found [here][3].

[3]: https://docs.docker.com/network/macvlan/

# Launching the Wi-SUN Docker Image

> Before you continue, you have to be aware that by default the image will try
> to detect existing networks. However this detection may fail (unusual network
> configuration, DHCP slow to answer, etc...). If no network is detected (and no
> command line option prevents it), this image advertises a network
> configuration. It may impact the other hosts on your network. See also [Can
> this image breaks my local network?](#can-this-image-breaks-my-local-network).

At this stage, you have:
  * A Docker image named `wisun-img`, running the Wi-SUN stack upper-layer and a
    communication protocol to control an EFR32 running the Wi-SUN RCP image.
  * A macvlan interface named `wisun-net`, allowing a Docker container to access
    one of the Raspberry Pi IP communication interfaces (by default `eth0`).

Before connecting the mainboard to your Linux host, start dmesg to monitor new
connected devices:

    dmesg -w

Next, connect the mainboard with the EFR32 radio board to the Raspberry Pi
through USB. Check the name of the new connected device in the dmesg output (by
default `/dev/ttyACM0`). If the device name is different, pass the correct
device name with `-d` (`--device`) in the following `docker run` command.

Launch your image using:

    docker run -ti --privileged --rm --network=wisun-net --name=wisun-vm wisun-img --flash -

> The `--flash -` option flashes the EFR32 with the default RCP image. More
> information in the [Using the JTAG Link](#using-the-jtag-link) section.

From now on, the Linux Wi-SUN border router should be running and you should be
able to connect Wi-SUN nodes to it.

Note that the Docker image accepts several options which you can list with:

    docker run -ti --privileged --rm wisun-img --help

You may want to open a shell into the container:

    docker exec -ti wisun-vm sh

> Advanced users may want to drop the `--rm` option and work with `docker
> start`, `docker stop`, etc...

# Using the JTAG Link

The Docker image is able to use the JTAG link available with the Silicon Labs
chips to configure the Wi-SUN parameters, flash the radio board or retrieve chip
traces.

You can start the container with `--flash -` to flash the connected board with
the built-in firmware. This firmware works with the EFR23xG12 radio boards
supporting a Wi-SUN sub-GHz RF band:

    docker run -ti --privileged --rm --network=wisun-net --name=wisun-vm wisun-img --flash -

> Replace "-" by another path to flash a custom firmware.

You can also use the JTAG link to retrieve traces from the radio board. You can
either pass `-T` to the `docker run` command or start the traces afterward with
`wisun-device-traces` command:

    docker exec -ti wisun-vm wisun-device-traces

# Using DBus interface

From 0.3.0, wsbrd offer a DBus interface to interact with the border router.
This image include [`busctl`][6] a generic tools to communicate with DBus. So,
user can run commands like:

    busctl introspect com.silabs.Wisun.BorderRouter /com/silabs/Wisun/BorderRouter

Obviously, this command only work from the inside the docker. So, from outside
the Docker, the full command is:

    docker exec wisun-vm busctl introspect com.silabs.Wisun.BorderRouter /com/silabs/Wisun/BorderRouter

> The formal specification of the DBus interface is not yet written, however
> most of the commands are self-described.

Here are a few examples of things you can using DBus.

Retrieve GAKs:

    busctl --user get-property com.silabs.Wisun.BorderRouter /com/silabs/Wisun/BorderRouter com.silabs.Wisun.BorderRouter Gaks

Monitor GAKs changes:

    busctl --match=type=signal --user monitor com.silabs.Wisun.BorderRouter

Revoke a EUI64:

    busctl --user call com.silabs.Wisun.BorderRouter /com/silabs/Wisun/BorderRouter com.silabs.Wisun.BorderRouter RevokeNode ay 8 0x01 0x23 0x34 0x56 0x78 0x9A 0xBC 0xDE

[6]: https://www.freedesktop.org/software/systemd/man/busctl.html

# Bugs and Limitations

## How to use a local copy of `wsbrd` sources

During the build, the Dockerfile checks if directory `wisun-br-linux` exists. In
this case, Dockerfile uses this directory instead of cloning the original
repository. So, you can do:

    git clone ssh://git@github.com/SiliconLabs/wisun-br-linux
    DOCKER_BUILDKIT=1 docker build --build-arg "GIT_DESCRIBE=$(git describe --tag --match v*)" -t wisun-img .


> Option `--ssh default` is no more necessary

You have to ensure the version of `wisun-br-linux` you provide is compatible
with this current Docker image.

## I want to use this architecture for production

By default, this Docker image uses a method called Neighbor Discovery Protocol
Proxy (NDP Proxy). It works with most IPv6 network topologies without touching
the network infrastructure.  However, it does not scale very well and you may
find limitations in corner cases. For production, use the subnet mode (aka
Prefix Delegation) or even better use DHCPv6-PD protocol (not presented in this
Docker image).

## I have no IPv6 network

This project does not aim to provide IPv6 connectivity. If your ISP does not
provide IPv6, you can either:

  - get equipment providing IPv6 through NAT64 or 6to4
  - get equipment advertising a site-local IPv6 prefix (eg. fd01::/64). You
    can do that using radvd with any standard Linux.

## The container does not detect my IPv6 network

The container relies on Router Advertisements. If your network uses DHCPv6, or
does not have Router Advertisement for any reason, the container won't detect
the network.

## Cannot reach (IPv4) Internet from the container

This happens when you use the macvlan driver. It is necessary to get an IP from
the DHCP server of the host network. Just add `-D` when you run the Docker image
to run a DHCP client:

    docker run -ti --privileged --rm --network=wisun-net --name=wisun-vm wisun-img -D

## I have restarted my Docker image and I can't ping my Wi-SUN device anymore

The proxy creates the necessary routes when it receives a Neighbor Solicitation.
Your host has probably cached this information. The easiest way to fix that is
to flush the neighbor information of your host with:

    ip -6 neigh flush dev eth0

Alternatively you can force a neighbor discovery on your Wi-SUN node:

    ndisc6 2a01:e35:2435:66a0:202:f7ff:fef0:0 eth0

## Wi-SUN can reach an outside network, but can't reach Docker host

It is a [limitation of the macvlan interface][5]. This situation is actually not
an error — it is the defined behavior of macvtap. Due to the way in which the
host's physical Ethernet is attached to the macvtap bridge, traffic into that
bridge from the guests that is forwarded to the physical interface cannot be
bounced back up to the host's IP stack. Additionally, traffic from the host's IP
stack that is sent to the physical interface cannot be bounced back up to the
macvtap bridge for forwarding to the guests.

There are several ways to work around the problem. The easiest way probably is
to use a secondary physical network interface exclusively for the guest.

    dhcpcd --release eth1
    docker network create -d macvlan -o parent=eth1 wisun-net
    ip link set dev eth1 up
    docker run -ti --privileged --rm --network=wisun-net --name=wisun-vm wisun-img

[5]: https://access.redhat.com/documentation/en-US/Red_Hat_Enterprise_Linux/6/html/Virtualization_Host_Configuration_and_Guest_Installation_Guide/App_Macvtap.html

## Unable to launch the container on my Windows workstation

This project has not yet been tested on Windows hosts. It seems it should work
as soon as you use Windows Subsystem for Linux (WSL2) and the USB-UART of the
Wi-SUN BR is handled by WSL2. In other words, you should see /dev/ttyUSB0 on
WSL2.

## I have reconnected the mainboard and nothing works

The Docker container does not (yet) support device hot-plugging. You have to
restart the Docker container if you unplug the mainboard with the EFR32 radio
board.

## When I try to ping from my Wi-SUN Device, the reply is transmitted after 5s of latency

When using the proxy, it takes a few seconds to establish connection the first
time an end device tries to access the outside. The problem is [ndppd does not
receive locally generated neighbor solicitation][6] (A). The system unlocks when
a solicitation comes from outside (B).

        tun0  2 1.806167960 2a01:e35:2435:66a0:20d:6fff:fe20:c096 → 2a00:1450:4007:809::200e ICMPv6 104 Echo (ping) request id=0x0001, seq=0, hop limit=63
        eth0  1 0.000000000 2a01:e35:2435:66a0:20d:6fff:fe20:c096 → 2a00:1450:4007:809::200e ICMPv6 118 Echo (ping) request id=0x0001, seq=0, hop limit=62
        eth0  2 0.007452561 2a00:1450:4007:809::200e → 2a01:e35:2435:66a0:20d:6fff:fe20:c096 ICMPv6 118 Echo (ping) reply id=0x0001, seq=0, hop limit=118 (request in 1)
    (A) eth0  3 0.007558306 fe80::42:acff:fe13:2 → ff02::1:ff20:c096 ICMPv6 86 Neighbor Solicitation for 2a01:e35:2435:66a0:20d:6fff:fe20:c096 from 02:42:ac:13:00:02
    (A) eth0  4 1.016063581 fe80::42:acff:fe13:2 → ff02::1:ff20:c096 ICMPv6 86 Neighbor Solicitation for 2a01:e35:2435:66a0:20d:6fff:fe20:c096 from 02:42:ac:13:00:02
    (A) eth0  5 2.039971376 fe80::42:acff:fe13:2 → ff02::1:ff20:c096 ICMPv6 86 Neighbor Solicitation for 2a01:e35:2435:66a0:20d:6fff:fe20:c096 from 02:42:ac:13:00:02
        eth0  6 3.060057826 2a01:e35:2435:66a0:42:acff:fe13:2 → 2a00:1450:4007:809::200e ICMPv6 166 Destination Unreachable (Address unreachable)
    (B) eth0  7 5.043717586 fe80::224:d4ff:fea3:4493 → 2a01:e35:2435:66a0:20d:6fff:fe20:c096 ICMPv6 86 Neighbor Solicitation for 2a01:e35:2435:66a0:20d:6fff:fe20:c096 from 00:24:d4:a3:44:93
        eth0  8 5.043761372 fe80::42:acff:fe13:2 → fe80::224:d4ff:fea3:4493 ICMPv6 174 Redirect
        eth0  9 5.043782371 fe80::42:acff:fe13:2 → ff02::1:ff20:c096 ICMPv6 86 Neighbor Solicitation for 2a01:e35:2435:66a0:20d:6fff:fe20:c096 from 02:42:ac:13:00:02
    (B) tun0  3 6.850276506 fe80::10c3:41bf:aa66:69a3 → ff02::1:ff20:c096 ICMPv6 72 Neighbor Solicitation for 2a01:e35:2435:66a0:20d:6fff:fe20:c096 from 00:00:00:00:00:00
        tun0  4 6.870837671 fe80::202:f7ff:fef0:0 → fe80::10c3:41bf:aa66:69a3 ICMPv6 72 Neighbor Advertisement 2a01:e35:2435:66a0:20d:6fff:fe20:c096 (sol) is at 00:02:f7:f0:00:00
        eth0 10 5.067321429 fe80::42:acff:fe13:2 → fe80::224:d4ff:fea3:4493 ICMPv6 86 Neighbor Advertisement 2a01:e35:2435:66a0:20d:6fff:fe20:c096 (rtr, sol) is at 02:42:ac:13:00:02

The subnet mode does not suffer from this limitation.

[6]: https://github.com/DanielAdolfsson/ndppd/issues/69

## Can this image break my local network?

When this image starts in `site_local` mode, it will advertise an IPv6 network
configuration. The other hosts on the local network will also receive this
configuration and will use it. From here, there are two situations:

  1. The hosts already have a valid IPv6 configuration. The new configuration
     will probably break the current one.

  2. The other hosts only use IPv4. IPv4 traffic won't be impacted. However,
     some DNS records may contain IPv4 and IPv6 fields. In this case, the
     applications may try to use the new IPv6 configuration. The hosts will
     forward IPv6 traffic to the Docker image. The image will return an error.
     Most of the time, the application will fall back to IPv4 and it will work.
     Else, the application will just stop here with a failure.


If you don't enforce `site_local`, the image will try to detect existing IPv6
networks. `site_local` will be started only if no IPv6 network has been
detected. The first scenario is unlikely to happen.

If you use `site_local` mode, we recommend using a dedicated network.

## I get `Error Connecting DP: cannot read IDR` when I try to flash

Check the power supply of the RF board (also check that RF board is correctly
present).

On the WTSK mainboard the switch labelled `BATT/USB/AEM` must be set
to `AEM`.

# Further improvements

* Replace radvd with a small RS/RA (Router Solicitation/Router Advertisement)
  proxy. nd-proxy.c seems to mostly do the job, but:
   1. For an unknown reason, it does not receive RS from tun0 and does not send
      RA to tun0 (while radvd is able to do that very well).
   2. It is written in C++.
* Provide an example of Prefix Delegation and DHCPv6-PD
* Docker does not yet support `--ipam-driver=dhcp --ipam-opt
  dhcp_interface=eth0`


CHANGES
v2.8
----

  - Bump wsbrd to v1.4
  - Bump built-in RCP to v1.2.0. Note this firmware still not work with EFR
    series 2 (EFRxG2x). For now, the docker image is still unable to detect the
    type of connected device.

v2.7
----

  - Bump wsbrd to v1.4
  - For the proxy mode, replace ndppd by the native support provided by wsbrd.
    This new way to handle neighbor proxy is far more reliable.

v2.6.1
------

  - Fix incompatibility with last update of github.com

v2.6
----

  - Include wsbrd_cli
  - Bump wsbrd to v1.3.3
  - Bump built-in RCP to v1.0.0

v2.5
----

  - Add dependencies for future wsbrd releaes
  - Simplify the way the device is flashed
  - Bump wsbrd to v1.2.0
  - Bump built-in RCP to v0.6.0

v2.4.4
------

  - In case the container was aborded, it might become unstarted. It was
    particulary annoying when the docker was started with
    --restart=unless-stopped.
  - Bump wsbrd to v1.1.0
  - Bump built-in RCP to v0.5.0

v2.4.3
------

  - Solve issues with Github authentication. Since the repository is now public,
    we can just use https transport.

v2.4.2
------

  - Bump wsbrd to v1.0.4

v2.4.1
------

  - Bump wsbrd to v1.0.1
  - Include mbedTLS (necessary for future wsbrd releases)

v2.4
----

  - Bump wsbrd to v1.0.0
  - Bump built-in RCP to v0.4.0
  - Include support for DBus. User can now run commands like:
        docker exec wisun-vm busctl introspect com.silabs.Wisun.BorderRouter /com/silabs/Wisun/BorderRouter
  - Remove tshark so, the size of the image greatly shrink (150MB -> 50MB)

v2.3.1
------

  - Allow to use a local copy of wisun-br-linux sources.
  - Fix error when launching DHCPv6-PD

v2.3
----

  - Update wsbrd version to v0.2.0. radvd is no more necessary now.
  - Start to use a configuration file for wsbrd. So, the command line used to
    run wsbrd is now simpler.
  - Do not advertise IPv6 network if an IPv4 network is detected
  - Fix support for -D
  - Improve DHCPv6-PD. unfortunatly, it is still experimental since correct
    implementation of RFC6204 is still rare on the market
  - Add progress bar when probing network
  - Fix range of valid Wi-SUN classes

v2.2
----

  - Rework documentation

v2.1
----

  - Bump wsbrd to v0.1.1 (no changes for end user)

v2.0
----

  - The Wi-SUN service now runs on the host
  - Show version numbers on start-up
  - Add way to flash the firmware from the docker image
  - Provide a default firmware
  - Add experimental support for DHCPv6-PD

v1.0
----

  - Initial release



init-container.sh
#!/bin/sh
# vim: set sw=4 expandtab:
#
# Copyright 2021, Silicon Labs
# SPDX-License-Identifier: Apache-2.0
# Main authors:
#     - Jérôme Pouiller <jerome.pouiller@silabs.com>
#
set +m

WSBRD_PID=-1
UART=/dev/ttyACM0

die()
{
    echo "$@" >&2
    exit 1
}

print_usage()
{
    cat << EOF >&2
Usage: $1 [OPTIONS] [WISUN_PARAMS] [MODE]

Setup the docker container to create a Wi-SUN Border Router.

Container options:
  -u, --uart=DEVICE   UART device to use (default: /dev/ttyACM0).
  -D, --dhcp          Configure IPv4 using DHCP. Use it if you rely on a
                      network interface with macvlan driver.
  -r, --advert-route  Advertise the new route on eth0. Only works with the subnet
                      mode. Most of the hosts won't accept le route unless the
                      parameter accept_ra_rt_info_max_plen is at least the size
                      of the advertised prefix size. You may use it if the
                      router of your network is not able to manage the new route
                      itself.
  -F, --flash=FW_PATH Flash radio board with FW_PATH. If FW_PATH is "-", flash
                      built-in firmware which should work with most of the
                      boards.
  -T, --chip-traces   Show traces from the chip.
  -s, --shell         Launch a shell on startup.
  -h, --help          Show this help.

Modes:
  local           The nodes will be only able to communicate with the docker
                  instance using a random site-local prefix.
  site_local      Advertise a random site-local prefix and run a proxy. Local
                  workstations will retrieve an IPv6 address allowing them to
                  communicate with Wi-SUN nodes.
  proxy           Re-use the local IPv6 prefix to configure Wi-SUN nodes.
  subnet [PREFIX] Use PREFIX to configure Wi-SUN nodes. PREFIX should come from
                  configuration of the parent router. If PREFIX is 'dhcp', get
                  prefix using DHCPv6-PD (experimental). If PREFIX is not
                  defined, generate a random site-local one.
  auto            Detect if a local IPv6 network is available and launch
                  \`local', \`site_local' or \`proxy' accordingly.

Note that random site-local prefixes are not routable (ie. you can't access
outside with these).

Wi-SUN parameters:
  -n, --ws-network=NAME    Set Wi-SUN network name (default: Wi-SUN)
  -d, --ws-domain=COUNTRY  Set Wi-SUN regulatory domain. Valid values: WW, EU,
                           NA (default), JP...
  -m, --ws-mode=VAL        Set operating mode. Valid values: 1a, 1b (default),
                           2a, 2b, 3, 4a, 4b and 5
  -c, --ws-class=VAL       Set operating class. Valid values: 1 (default), 2, 3
                           or 4
  -S, --ws-size=SIZE       Optimize network timings considering the number of
                           expected nodes on the network. Valid values: CERT
                           (development and certification), S (< 100, default),
                           M (100-800), L (800-2500), XL (> 2500)
  -K, --ws-key=FILE        Private key (default: br_key.pem)
  -C, --ws-cert=FILE       Certificate for the key (default: br_cert.pem)
  -A, --ws-authority=FILE  Certificate of the authority (CA) (default:
                           ca_cert.pem)

By default, this container uses embedded test certficates (located in
/usr/local/share/wsbrd/examples/). If you provide your owns, don't forget to map
the files into the container (passing option -v to \`docker run').

Examples:

  Provide minimal infrastructure to configure Wi-SUN device through a Border
  Router connected on /dev/ttyUSB0:

    $1 -d /dev/ttyUSB0 local

  Parent router is correctly configured to delegate prefix
  2a01:e35:2435:66a1::/64 to my docker container:

    $1 subnet 2a01:e35:2435:66a1::1/64

  You want to test prefix delegation with a random prefix:

    other_host> sysctl net.ipv6.conf.eth0.accept_ra_rt_info_max_plen=128
    this_host> $1 subnet -r
EOF
    [ "$2" ] && exit $2
}

get_random_prefix()
{
    N1=$(dd bs=1 count=1 if=/dev/urandom 2> /dev/null | od -A n -t x1)
    N2=$(dd bs=2 count=1 if=/dev/urandom 2> /dev/null | od -A n -t x2)
    N3=$(dd bs=2 count=1 if=/dev/urandom 2> /dev/null | od -A n -t x2)
    N4=$(dd bs=2 count=1 if=/dev/urandom 2> /dev/null | od -A n -t x2)
    echo $N1:$N2:$N3:$N4 | tr -d ' '
}

check_privilege()
{
    ip link add dummy0 type dummy 2> /dev/null || \
        die "Not enough privilege to run (missing --privileged?)"
    ip link delete dummy0
}

launch_dhcp4()
{
    if [ -s /run/dhcp/dhclient.pid ]; then
        [ -e /proc/$(cat /run/dhcp/dhclient.pid) ] && return
    fi
    umount /etc/resolv.conf
    ip -4 route flush dev eth0 default
    dhclient -nw eth0
}

launch_dhcp6()
{
    if [ -s /run/dhcp/dhclient6.pid ]; then
        [ -e /proc/$(cat /run/dhcp/dhclient6.pid) ] && return
    fi
    dhclient -nw -P --prefix-len-hint 64 eth0
}

launch_dbus()
{
    # Clean up possible previous instance
    rm -f /var/run/dbus.pid
    dbus-daemon --system
}

launch_wsbrd()
{
    IPV6_NET=$1
    [ -e "$UART" ] || die "Failed to detect $UART"

    echo " ---> [1mLaunch wsbrd[0m"
    echo "Configuration file:"
    sed -e 's/#.*//' -e '/^ *$/d' -e 's/^/    /' /etc/wsbrd.conf
    echo "Command line:"
    echo "    wsbrd -u $UART -F /etc/wsbrd.conf$WSBRD_ARGS -o ipv6_prefix=$IPV6_NET --network=\"$WS_NETWORK\""
    wsbrd -u $UART -F /etc/wsbrd.conf$WSBRD_ARGS -o ipv6_prefix=$IPV6_NET --network="$WS_NETWORK" &
    WSBRD_PID=$!

    for i in $(seq 100); do
        ip -6 addr show scope global | grep -q tun0 && break
        sleep 0.2
    done
}

generate_radvd_conf()
{
    IPV6_NET=$1
    EXT_BEHAVIOR=$2

    case "$EXT_BEHAVIOR" in
        adv_prefix)
            cat << EOF >> /etc/radvd.conf
interface eth0 {
    AdvSendAdvert on;
    AdvDefaultLifetime 0;
    prefix $IPV6_NET { };
};
EOF
            ;;
        adv_route)
            cat << EOF >> /etc/radvd.conf
interface eth0 {
    AdvSendAdvert on;
    AdvDefaultLifetime 0;
    route $IPV6_NET {
        AdvRouteLifetime 1800;
    };
};
EOF
            ;;
        "")
            ;;
        *)
            die "internal error: unknown options: $EXT_BEHAVIOR"
            ;;
    esac
}

launch_radvd()
{
    IPV6_NET=$1
    EXT_BEHAVIOR=$2

    echo " ---> [1mLaunch radvd on $IPV6_NET[0m"
    generate_radvd_conf $IPV6_NET $EXT_BEHAVIOR
    radvd --logmethod stderr
}

launch_last_process()
{
    echo " ---> [1mResult of 'ip -6 addr':[0m"
    ip -6 addr
    if [ "$LAUNCH_SHELL" ]; then
        if tty > /dev/nul 2> /dev/null; then
            set -m
        else
            echo "Cannot get tty (missing -t in docker command line?)"
        fi
        echo " ---> [1mLaunch sh[0m"
        exec sh
    elif [ "$LAUNCH_TRACES" ]; then
        echo " ---> [1mLaunch wisun-device-traces[0m"
        exec wisun-device-traces
    else
        wait $WSBRD_PID
        echo " ---> [1mWi-SUN border router has disappeared[0m"
    fi
}

flash_firmware()
{
    echo " ---> [1mFlash firmware[0m"
    [ -e "$FIRMWARE" ] || die "'$FIRMWARE' not found (missing -v in docker command?)"
    openocd -f board/efm32.cfg -c "program $FIRMWARE 0 reset exit"
    echo "You may have to hard-reset the board if it hangs"
}

run_proxy()
{
    sysctl -q net.ipv6.conf.all.forwarding=1

    for i in $(seq 10); do
        ip -6 addr show scope global | grep -q eth0 && break
        sleep 0.2
    done
    IPV6_NET=$(rdisc6 -r 10 -w 400 -q -1 eth0)
    [ "$IPV6_NET" ] || die "Failed to get IPv6 address"

    WSBRD_ARGS="$WSBRD_ARGS -o neighbor_proxy=eth0"
    launch_wsbrd $IPV6_NET
    launch_last_process
}

run_site_local()
{
    sysctl -q net.ipv6.conf.all.forwarding=1

    IPV6_NET=fd$(get_random_prefix)::/64
    launch_radvd $IPV6_NET adv_prefix
    WSBRD_ARGS="$WSBRD_ARGS -o neighbor_proxy=eth0"
    launch_wsbrd $IPV6_NET
    launch_last_process
}

run_local()
{
    IPV6_NET=fd$(get_random_prefix)::/64
    launch_wsbrd $IPV6_NET
    launch_last_process
}

run_subnet()
{
    sysctl -q net.ipv6.conf.all.forwarding=1

    IPV6_NET=${1:-fd$(get_random_prefix)::/64}
    if [ "$ADVERT_ROUTE" ]; then
        launch_radvd $IPV6_NET adv_route
    fi
    launch_wsbrd $IPV6_NET
    launch_last_process
}

run_dhcpv6pd()
{
    sysctl -q net.ipv6.conf.all.forwarding=1

    # Should comply with RFC6204
    launch_dhcp6
    printf "Wait for DHCP reply"
    for i in $(seq 20); do
        [ -e /tmp/dhcpv6pd.lease ] && break
        printf "."
        sleep 0.2
    done
    printf "\n"
    [ -e /tmp/dhcpv6pd.lease ] || die "Can't get prefix delegation"

    IPV6_PREFIX_LEN="$(cat /tmp/dhcpv6pd.lease | sed 's:.*/::')"
    IPV6_PREFIX="$(cat /tmp/dhcpv6pd.lease | sed 's:/.*::')"
    [ -n "$IPV6_PREFIX_LEN" ] || die "Can't get prefix delegation"
    [ -n "$IPV6_PREFIX" ] || die "Can't get prefix delegation"
    [ "$IPV6_PREFIX_LEN" -le 64 ] || die "IPv6 prefix must be at least /64"
    IPV6_NET=$IPV6_PREFIX/64
    if [ "$ADVERT_ROUTE" ]; then
        launch_radvd $IPV6_NET adv_route
    fi
    launch_wsbrd $IPV6_NET
    launch_last_process
}

run_auto()
{
    HAVE_IPV4=
    HAVE_IPV6=
    launch_dhcp4
    printf "Probe network"
    for i in $(seq 20); do
        ip -6 addr show scope global | grep -q eth0 && HAVE_IPV6=1 && break
        [ -s /var/lib/dhcp/dhclient.leases ] && HAVE_IPV4=1
        sleep 0.2
        printf "."
    done
    printf "\n"
    if [ "$HAVE_IPV6" ]; then
        echo " ---> [1mFound IPv6 network (launch proxy mode)[0m"
        run_proxy
    elif [ "$HAVE_IPV4" ]; then
        echo " ---> [1mFound IPv4 network (launch local mode)[0m"
        run_local
    else
        echo " ---> [1mNo network found (launch site_local mode)[0m"
        run_site_local
    fi
}

cat /etc/issue

WS_NETWORK="Wi-SUN"
OPTS=$(getopt -l shell,chip-traces,dhcp,device:,uart:,advert-route,flash:,ws-network:,ws-domain:,ws-mode:,ws-class:,ws-size:,ws-key:,ws-cert:,ws-authority:,help -- sTDu:rF:n:d:m:c:S:K:C:A:h "$@") || exit 1
eval set -- "$OPTS"
while true; do
    case "$1" in
        -s|--shell)
            LAUNCH_SHELL=1
            shift 1
            ;;
        -T|--chip-traces)
            LAUNCH_TRACES=1
            shift 1
            ;;
        -D|--dhcp)
            LAUNCH_DHCP4=1
            shift 1
            ;;
        -u|--uart|--device)
            UART=$2
            shift 2
            ;;
        -r|--advert-route)
            ADVERT_ROUTE=1
            shift 2
            ;;
        -F|--flash)
            if [ -z "$2" -o "$2" == "-" ]; then
                FIRMWARE=/firmware-winsun-rcp.s37
            else
                FIRMWARE=$2
            fi
            shift 2
            ;;
        -n|--ws-network)
            WS_NETWORK="$2"
            shift 2
            ;;
        -d|--ws-domain)
            WSBRD_ARGS="$WSBRD_ARGS -o domain=$2"
            shift 2
            ;;
        -m|--ws-mode)
            WSBRD_ARGS="$WSBRD_ARGS -o mode=$2"
            shift 2
            ;;
        -c|--ws-class)
            WSBRD_ARGS="$WSBRD_ARGS -o class=$2"
            shift 2
            ;;
        -S|--ws-size)
            WSBRD_ARGS="$WSBRD_ARGS -o size=$2"
            shift 2
            ;;
        -K|--ws-key)
            WSBRD_ARGS="$WSBRD_ARGS -o key='$2'"
            shift 2
            ;;
        -C|--ws-cert)
            WSBRD_ARGS="$WSBRD_ARGS -o certificate='$2'"
            shift 2
            ;;
        -A|--ws-authority)
            WSBRD_ARGS="$WSBRD_ARGS -o authority='$2'"
            shift 2
            ;;
        -h|--help)
            print_usage $0
            exit 0
            ;;
        --)
            shift
            break
            ;;
    esac
done
[ "$LAUNCH_SHELL" -a "$LAUNCH_TRACES" ] && die "--shell and --chip-traces are exclusive"

check_privilege

# If accept_ra=1, the default route is dropped when IP forwarding is enabled
sysctl -q net.ipv6.conf.eth0.accept_ra=2
# Docker disables IPv6 by default, we need to reactivate it
sysctl -q net.ipv6.conf.all.disable_ipv6=0
launch_dbus
[ "$LAUNCH_DHCP4" ] && launch_dhcp4
[ "$FIRMWARE" ] && flash_firmware

case "$1" in
    auto|"")
        run_auto
        ;;
    site_local)
        run_site_local
        ;;
    local)
        run_local
        ;;
    proxy)
        run_proxy
        ;;
    subnet)
        if [ "$2" = dhcp -o "$2" = DHCP ]; then
            run_dhcpv6pd
        else
            run_subnet $2
        fi
        ;;
    *)
        print_usage $0
        exit 1
esac
